### ISSUE TITLE
Key to smithy doesn't spawn 

### ISSUE THREAD

**Issue by AmProsius** [AmProsius/gothic-1-community-patch#47]
**Describe the bug**
The door to the smithy in the Old Camp can't be unlocked for the entire game.

**Expected behavior**
The door to the smithy in the Old Camp can now be unlocked. The player can obtain the key throughout the game.

**Additional context**
It seems that the developers forgot to place the key in the world, so the placement of the key is chosen by the G1CP team.

**Comment by catalinstoian** [comment]
Either in
GRD_211_Skip.d

```
	CreateInvItem(self,ITKEY_OB_SMITH_01);

```
or

can be done via ZS_Dead.d or B_Kapitelwechsel.d (if we want the key to be available only after Ch.4)

**Comment by szapp** [comment]
I vote for adding it immediately. This way, the patch will also take effect, if it is installed for a later game save (say, after Chapter 4).

**Comment by AmProsius** [comment]
I agree for the exact same reason.

**Comment by AmProsius** [comment]
As discussed the key should be spawned in Skip's inventory in `B_Story_ExploreSunkenTower` when the dungeon key is spawned in Gomez' inventory.

**Comment by AmProsius** [comment]
https://github.com/AmProsius/gothic-1-community-patch/blob/b8996cb83c3157b047d6109a3b203c357bc9bebb/scriptbase/_work/Data/Scripts/Content/Story/Chapters/B_Story_ExploreSunkenTower.d#L68-L70

changed to

```d
    // Schlüssel um die Tür zum Kerker zu öffnen
    CreateInvItem ( Gomez, DungeonKey );

    // Schlüssel, um die Tür hinter der Schmiede zu öffnen
    CreateInvItem ( Skip, ItKe_OB_Smith_01 );
```

**Comment by szapp** [comment]
A question about this: Given that Skip's faith is not fully decided yet, does he still offer to trade at this time of the game? This way it would be nice for more "self-defense" oriented players to obtain the key without attacking Skip - if we decide to leave him neutral.

**Comment by AmProsius** [comment]
> A question about this: Given that Skip's faith is not fully decided yet, does he still offer to trade at this time of the game?

I will have to check that.

**Comment by szapp** [comment]
After giving it a bit more thought, I think this fix is very problematic. 

Say we make the key available - regardless of how we do it - we would have to remove it from the player’s inventory on saving the game (to follow the idea of making the patch fully removable without leaving a trace). Consequently, we’d have to re-insert the key when loading a game.

A big question follows: under what circumstances should the key be re-added to the inventory? E.g. chapter >= 4 and Gomez dead. But what if the player had the key but dropped it in a chest or the world, because they don’t like keys flooding their inventory? We would generate endless amounts of keys every time the game is loaded.

An alternative that comes to mind is, to go forward with fixing the door to require the correct key instance, but never insert the key. Instead just programmatically unlock the door from chapter 5 onward. That might, however, go beyond a “fix” for a missing key. Any thoughts or suggestions?

**Comment by szapp** [comment]
I think we might not have to remove the item from the inventory on saving. It’s an item of the original game (we are not modifying it). It is therefore comparable to #23 where we merely unlock the possibility to acquire an existing item.

The difference here is that when placing the key in the inventory of Skip or Gomez, is that the opportunity of acquiring the key vanishes as soon as they are killed and fully looted (as the body disappears). It would be easier if we would spawn the key at a location (around the smithy somewhere) instead of in an NPC inventory. On every load we could check:
1. Is it chapter 5 or higher?
2. Is the door still locked?
3. Does the key exist somewhere in the world?
3. Does the player have the key in the inventory?
Before spawning it at the specific location.

**Comment by AmProsius** [comment]
> The difference here is that when placing the key in the inventory of Skip or Gomez, is that the opportunity of acquiring the key vanishes as soon as they are killed and fully looted (as the body disappears).

Gomez is immortal until the keys are placed in his inventory, so I don't think that would be an issue. Skip on the other hand might be an issue, I agree.
I don't really like the idea of placing the key anywhere in the world.

**Comment by szapp** [comment]
> Gomez is immortal until the keys are placed in his inventory, so I don't think that would be an issue.

What I meant is that if the player has already gotten so far in the game (killed Gomez in chapter 5) before applying the patch, they will have no chance to acquire the key. As soon as Gomez is dead, looted and despawned, the patch cannot add the key to the game and the door remains inaccessible - the patch would no longer follow the idea of being able to install at any time. I would like to avoid that. 

The key could also be spawned in a chest (if not the world), or be available in the dungeons that are only accessible with the key obtained from Gomez.

**Comment by AmProsius** [comment]
> The key could also be spawned in a chest (if not the world), or be available in the dungeons that are only accessible with the key obtained from Gomez.

I like the idea with the chest, but I think we'd have the same problems as with an NPC's inventory. If the chest was looted before, the player won't return to it. The only possibility I see would be to place the key in a chest that only comes available after chapter 4. 

After giving it a bit more thought myself, I'd rather place the key somewhere in Stone's cell. The player than could obtain the key after freeing him.

**Comment by szapp** [comment]
> I like the idea with the chest, but I think we'd have the same problems as with an NPC's inventory. If the chest was looted before, the player won't return to it.

It’s not quite the same problem. The player might have already looted the chest, but they at least have the opportunity to go back and retrieve the key, whereas if Gomez is gone, there is no way to get the key.

Regardless of that, I agree that a chest is not a good option, because a mod might have removed the chest or moved it enough for us not to be able to identify it. That would not allow us to place the key in it. 

> After giving it a bit more thought myself, I'd rather place the key somewhere in Stone's cell (like you mentioned above). The player than could obtain the key after freeing him.

That sounds like a good alternative. I would suggest to place the key there from two locations in the scripts:
1. In `B_Story_ExploreSunkenTower` as suggest above
2. And on loading the game given the following conditions are all true:
    - It’s chapter 5 or higher
    - The door is identified and still locked
    - The key is not in the player’s inventory
    - The key is nowhere in the world

**Comment by AmProsius** [comment]
> It’s chapter 5 or higher

I think it's actually chapter 4 when the Old Camp turns against the player.

> The key is nowhere in the world

Does that also include chests? Then I agree, because the player may have stored the key.

**Comment by szapp** [comment]
> I think it's actually chapter 4 when the Old Camp turns against the player.

Ah, okay, I didn’t really remember when what happens during the higher chapters. It should be *after* `B_Story_ExploreSunkenTower`.

> > The key is nowhere in the world
> 
> Does that also include chests? Then I agree, because the player may have stored the key.

No, but you’re right that should be checked too. That might increase loading time, however. Maybe I can think of a more direct way of checking if the key exists in general.

**Comment by szapp** [comment]
### Conditions under which to create the key  
With `keyId = MEM_FindParserSymbol("ItKe_OB_Smith_01")`
1. The variable `ExploreSunkenTower` exists and is `TRUE`.  
    `if (Ninja_G1CP_GetIntSymbol("ExploreSunkenTower", FALSE))` (function does not exist yet)
2. The key instance exists in the scripts  
    `if (keyId != -1)`
3. The door [is identified](https://github.com/AmProsius/gothic-1-community-patch/blob/20118f9dbf64495c877d110ddc6a4241ea52d3f6/src/Ninja/G1CP/Content/Fixes/Gamesave/fix046_SmithDoor.d#L20-L23) (i.e. PC is currently in the correct world), see #46.
    - It's locked  
    `if (mob._oCMobLockable_bitfield | oCMobLockable_bitfield_locked)`
    - It has the correct key instance  
    `if (Hlp_StrCmp(mob._oCMobLockable_keyInstance, "ItKe_OB_Smith_01"))`
4. Check if target waypoint exists, candidate: "OCC_Cellar_Back_Left_Cell".  
    `if (Ninja_G1CP_GetWaypoint("OCC_Cellar_Back_Left_Cell"))` (function does not exist yet)
5. The key is nowhere in the world. (This will be quite challenging to implement.)
    - Not physically in the world  
    `if (!MEM_SearchVobByName("ItKe_OB_Smith_01"))`
    - Not in the PC or any NPC inventory  
    `if (!Npc_HasItems(slf, keyId))` for all `slf` in `MEM_World.voblist_npcs`
    - Not in any container  
    `if (!Mob_HasItems(mob, keyId))` for all `mob` in `MEM_World.voblist`
    - *These checks are quite **performance demanding** and might have to be **implemented with machine code**!*
    - *Double check if the inventory of NPC's is packed (like the PC's) and same for containers. If they are not packed it might be much more easy without looping over all VOBs in the world and without the need for machine code.*

Insert the key if all conditions above are met (suggested waypoint in Stone's dungeon cell):  
`Wld_InsertItem(keyId, "OCC_Cellar_Back_Left_Cell")`

**Comment by szapp** [comment]
If we were to merge #46 and #47 with out own key instance, it could look like this:

### Apply
```
1. Is door valid (identified,
   locked, wrong key instance)
     │             │
     │ True        │ False
     │             └──────────────────────────────────────────┐
     V                                                        │
2. Does the G1CP key already exist in the world               │
     │             │                                          │
     │ True        │ False                                    │
     │             │                                          │
     │             V                                          │
     │        A. Can key be inserted?                         │
     │           (waypoint exists)                            │
     │             │            │                             │
     │             │ True       │ False                       │
     │             │            └──────────────────────────┐  │
     │             V                                       │  │
     │        B. Is story progressed enough?               │  │
     │           (ExploreSunkenTower == True)              │  │
     │             │            │                          │  │
     │             │ True       │ False                    │  │
     │             │            │                          │  │
     │             │            V                          │  │
     │             │     I.  Is story variable set in      │  │
     │             │         expected function?            │  │
     │             │         (ExploreSunkenTower in        │  │
     │             │          B_Story_ExploreSunkenTower)  │  │
     │             │            │           │              │  │
     │             │            │ True      │ False        │  │
     │             │            │           └───────────┐  │  │
     │             │            V                       │  │  │
     │             │     II. Hook after the story       │  │  │
     │             │         function to insert the     │  │  │
     │             │         key then                   │  │  │
     │             │            │                       │  │  │
     │    ┌─────── │ ───────────┘                       │  │  │
     │    │        │                                    │  │  │
     │    │        │                                    │  │  │
     │    │        V                                    │  │  │
     │    │   C. Insert the key                         │  │  │
     │    │      (in Stone's cell)                      │  │  │
     │    │        │           │                        │  │  │
     │    │        │ Success   │ Failed (unlikely)      │  │  │
     │    │        │           └─────────────────────┐  │  │  │
     │    │        │                                 │  │  │  │
     V    V        V                                 │  │  │  │
3. Adjust the key instance                           │  │  │  │
   for the door                                      │  │  │  │
        │                                            │  │  │  │
        │                                            │  │  │  │
        V                                            V  V  V  V
    return True                                     return False
    (fix applied)                              (fix not applied)
```

### Revert
```
1. Is the fix currently applied
     │
     │ True
     V
2. Is door valid (identified,
   locked, wrong key instance)
     │
     │ True
     V
3. Adjust the key instance
```

That would mean, the fix (i.e. adjusting the door) would only be applied if
1. the door has the wrong key instance
2. the story variables and functions exist as expected

Would the story be changed somewhat (the variables renamed) the entire fix would not be applied.

**Comment by AmProsius** [comment]
> That would mean, the fix (i.e. adjusting the door) would only be applied if
>
> 1. the door has the incorrect key instance
> 2. the story variables and functions exist as expected

I would add that the original key still has the original instance name as many mods fixed this issue by renaming the instance rather than the property of the door.

**Comment by szapp** [comment]
> > That would mean, the fix (i.e. adjusting the door) would only be applied if
> > 
> > 1. the door has the incorrect key instance
> > 2. the story variables and functions exist as expected
> 
> I would add that the original key still has the original instance name as many mods fixed this issue by renaming the instance rather than the property of the door.
> In this case, that would lead to two keys from which only one works (the one from the G1CP).

True, I will add a check for that in this flow chart (that check is already in #167). Give me a few minutes and check above if that's what you meant.

**Comment by szapp** [comment]
Duplicate of #46.

If it hasn't been deleted, the implementation discussed here can be found in the branch [bug047](../tree/bug047), specifically in commit 1446f15.
